File: src/__tests__/truncation-recovery.test.ts
/**
 * Feature: spec/features/truncated-tool-call-recovery-auto-chunk-large-writes-and-retry-on-max-tokens.feature
 *
 * PROV-040: Truncated tool call recovery — auto-chunk large writes and retry on max_tokens
 *
 * Integration tests that verify the truncation detection and recovery message contract
 * using fixture data. These tests validate the same behavior the Rust implementation
 * enforces, ensuring TypeScript/TUI consumers handle truncation correctly.
 *
 * NO MOCKS — Uses fixture data and real string matching.
 */

import { describe, it, expect } from 'vitest';

// =============================================================================
// Fixture data: realistic error strings from each provider
// =============================================================================

/** Truncation error fixtures matching the PROV-039 enriched error format */
const truncationErrorFixtures = {
  anthropicWrite: [
    'Streaming error: ResponseError: Tool call truncated due to output token limit.',
    "Tool 'Write' received incomplete JSON arguments.",
    'The model hit max_tokens while generating the tool call.',
    'Partial arguments: {"file_path": "/tmp/prov039-large-write-test.txt"',
  ].join(' '),

  anthropicBash: [
    'Tool call truncated due to output token limit.',
    "Tool 'Bash' received incomplete JSON arguments.",
    'The model hit max_tokens while generating the tool call.',
    'Partial arguments: {"command": "cat << \'HEREDOC_EOF\'',
  ].join(' '),

  openaiEdit: [
    'Tool call truncated due to output token limit.',
    "Tool 'Edit' received incomplete JSON arguments.",
    'Partial arguments: {"file_path": "/src/main.rs", "old_string": "fn',
  ].join(' '),

  geminiWrite: [
    'Tool call truncated due to output token limit.',
    "Tool 'Write' received incomplete JSON arguments.",
    'Partial arguments: {"file_path": "/tmp/output.json", "content": "[{',
  ].join(' '),
};

/** Non-truncation error fixtures that must NOT trigger recovery */
const nonTruncationErrorFixtures = {
  networkTimeout: 'Network timeout after 30000ms',
  authFailure: 'Authentication failed: Invalid API key',
  rateLimited: 'Rate limit exceeded. Please retry after 60 seconds.',
  promptTooLong: 'prompt is too long: 209834 tokens > 200000 maximum',
  contextExceeded: 'context_length_exceeded: Request too large',
  thinkingBudget:
    '{"type":"invalid_request_error","message":"`max_tokens` must be greater than `thinking.budget_tokens`"}',
  genericServerError: 'Internal server error (500)',
  sseError: 'SSE Error: connection reset by peer',
  textTruncation: 'Response truncated: model hit max_tokens output limit',
  partialSubstring: 'Tool call truncated',
  partialSubstring2: 'output token limit exceeded',
};

// =============================================================================
// Detection contract: matches the Rust is_truncated_tool_call_error()
// =============================================================================

/**
 * TypeScript-side detection function matching the Rust implementation.
 * This validates the contract: "contains 'Tool call truncated due to output token limit'"
 *
 * SYNC NOTE: This sentinel string is produced by rig-core's anthropic/streaming.rs
 * (PROV-039, line 443) and is also checked by Rust's is_truncated_tool_call_error()
 * in stream_loop.rs. If the sentinel changes, both this TS function AND the Rust
 * function must be updated. The Rust tests in truncation_recovery_test.rs test the
 * same string, so a desync would be caught by cargo test failing.
 */
function isTruncatedToolCallError(errorStr: string): boolean {
  return errorStr.includes('Tool call truncated due to output token limit');
}

/**
 * Extract tool name from the truncation error message.
 * Matches the Rust build_truncation_recovery_message() extraction logic.
 */
function extractToolName(errorStr: string): string {
  const match = errorStr.match(/Tool '([^']+)'/);
  return match ? match[1] : 'unknown';
}

/**
 * Extract partial arguments from the truncation error message.
 * Matches the Rust build_truncation_recovery_message() extraction logic.
 */
function extractPartialArgs(errorStr: string): string {
  const parts = errorStr.split('Partial arguments: ');
  return parts.length > 1 ? parts[1] : '(not available)';
}

// =============================================================================
// Tests
// =============================================================================

describe('Feature: Truncated tool call recovery', () => {
  describe('Scenario: Truncated tool call error includes structured recovery instruction', () => {
    it('should detect Anthropic Write truncation error', () => {
      // @step Given the agent is streaming a response from any provider
      const error = truncationErrorFixtures.anthropicWrite;

      // @step And the model attempts a Write tool call with content exceeding the output token limit
      // @step When the tool call is truncated due to max_tokens
      const detected = isTruncatedToolCallError(error);

      // @step Then the error message contains a structured recovery instruction
      expect(detected).toBe(true);

      // @step And the recovery instruction includes the truncated tool name
      expect(extractToolName(error)).toBe('Write');

      // @step And the recovery instruction includes the partial arguments that were received
      expect(extractPartialArgs(error)).toContain(
        '/tmp/prov039-large-write-test.txt'
      );
    });

    it('should extract tool name and partial args for all provider errors', () => {
      // @step Given the agent is streaming a response from any provider
      const fixtures = [
        {
          error: truncationErrorFixtures.anthropicWrite,
          tool: 'Write',
          argFragment: '/tmp/prov039',
        },
        {
          error: truncationErrorFixtures.anthropicBash,
          tool: 'Bash',
          argFragment: 'HEREDOC_EOF',
        },
        {
          error: truncationErrorFixtures.openaiEdit,
          tool: 'Edit',
          argFragment: '/src/main.rs',
        },
        {
          error: truncationErrorFixtures.geminiWrite,
          tool: 'Write',
          argFragment: '/tmp/output.json',
        },
      ];

      for (const fixture of fixtures) {
        // @step When the tool call is truncated due to max_tokens
        expect(isTruncatedToolCallError(fixture.error)).toBe(true);

        // @step And the recovery instruction includes the truncated tool name
        expect(extractToolName(fixture.error)).toBe(fixture.tool);

        // @step And the recovery instruction includes the partial arguments that were received
        expect(extractPartialArgs(fixture.error)).toContain(
          fixture.argFragment
        );
      }
    });
  });

  describe('Scenario: Retry budget prevents infinite truncation retry loops', () => {
    it('should allow exactly MAX_TRUNCATION_RETRIES before exhaustion', () => {
      // @step Given the agent is streaming a response from any provider
      // @step And the truncation retry budget is set to 2
      const MAX_RETRIES = 2;
      let retryCount = 0;
      const error = truncationErrorFixtures.anthropicWrite;

      // @step When the model hits max_tokens on the first tool call attempt
      expect(isTruncatedToolCallError(error)).toBe(true);
      retryCount++;

      // @step Then the recovery instruction is sent and a retry stream is started
      expect(retryCount).toBeLessThanOrEqual(MAX_RETRIES);

      // @step When the model hits max_tokens on the second tool call attempt
      retryCount++;
      expect(retryCount).toBeLessThanOrEqual(MAX_RETRIES);

      // Third attempt — budget exhausted
      retryCount++;

      // @step Then the retry budget is exhausted
      expect(retryCount).toBeGreaterThan(MAX_RETRIES);

      // @step And the error is reported to the user as a non-recoverable truncation failure
      // @step And the stream loop terminates without starting another retry
      // (In Rust: the else branch emits the budget-exhausted error and returns Err)
    });
  });

  describe('Scenario: Normal completion is unaffected by truncation recovery logic', () => {
    it('should not detect non-truncation errors as truncation', () => {
      // @step Given the agent is streaming a response from any provider
      // @step And the model completes a tool call normally with stop_reason end_turn
      const errors = Object.values(nonTruncationErrorFixtures);

      for (const error of errors) {
        // @step When the stream completes
        // @step Then no recovery instruction is injected
        expect(isTruncatedToolCallError(error)).toBe(false);
      }

      // @step And the truncation retry counter remains at zero
      // @step And the behavior is identical to pre-PROV-040 baseline
    });
  });

  describe('Scenario: Text-only truncation does not trigger tool call recovery', () => {
    it('should not detect text truncation warning as tool call truncation', () => {
      // @step Given the agent is streaming a response from any provider
      // @step And the model hits max_tokens during a text-only response with no tool call
      const textTruncation =
        'Response truncated: model hit max_tokens output limit';

      // @step When the stream completes with stop_reason max_tokens
      // @step Then the existing PROV-039 truncation warning is displayed
      // @step And no tool call recovery instruction is injected
      expect(isTruncatedToolCallError(textTruncation)).toBe(false);
    });
  });

  describe('Scenario: Truncation recovery is provider-agnostic', () => {
    it('should detect truncation errors from all providers identically', () => {
      // @step Given the truncation detection relies on the error message string from PROV-039
      const providerErrors = [
        truncationErrorFixtures.anthropicWrite,
        truncationErrorFixtures.openaiEdit,
        truncationErrorFixtures.geminiWrite,
      ];

      // @step When a truncation error containing "Tool call truncated due to output token limit" is received
      for (const error of providerErrors) {
        // @step Then the same recovery logic fires regardless of whether the provider is Anthropic, OpenAI, or Gemini
        expect(isTruncatedToolCallError(error)).toBe(true);
      }

      // @step And the recovery instruction content is identical across all providers
      // All use the same detection string and recovery message builder
      const toolNames = providerErrors.map(e => extractToolName(e));
      // Each extracts a valid tool name (not 'unknown')
      for (const name of toolNames) {
        expect(name).not.toBe('unknown');
      }
    });
  });

  describe('Edge cases', () => {
    it('should not match empty string', () => {
      expect(isTruncatedToolCallError('')).toBe(false);
    });

    it('should not match partial substrings', () => {
      expect(isTruncatedToolCallError('Tool call truncated')).toBe(false);
      expect(isTruncatedToolCallError('output token limit')).toBe(false);
      expect(isTruncatedToolCallError('truncated due to')).toBe(false);
    });

    it('should handle tool name extraction for unknown format', () => {
      const weirdError =
        'Tool call truncated due to output token limit. No standard format here.';
      expect(isTruncatedToolCallError(weirdError)).toBe(true);
      expect(extractToolName(weirdError)).toBe('unknown');
    });

    it('should handle partial args extraction when missing', () => {
      const noArgs =
        "Tool call truncated due to output token limit. Tool 'Write' received incomplete JSON.";
      expect(extractPartialArgs(noArgs)).toBe('(not available)');
    });
  });
});

File: codelet/cli/tests/truncation_recovery_test.rs
#![allow(clippy::unwrap_used, clippy::expect_used, clippy::panic, clippy::uninlined_format_args)]
//! Feature: spec/features/truncated-tool-call-recovery-auto-chunk-large-writes-and-retry-on-max-tokens.feature
//!
//! PROV-040: Truncated tool call recovery — auto-chunk large writes and retry on max_tokens
//!
//! Tests for truncated tool call detection, recovery message generation, and
//! retry budget logic. These tests import the REAL production functions from
//! codelet_cli::interactive — no copies, no mocks.

use codelet_cli::interactive::{
    build_truncation_budget_exhausted_message, build_truncation_recovery_message,
    is_truncated_tool_call_error, MAX_TRUNCATION_RETRIES,
};

// =============================================================================
// Scenario: Truncated tool call error includes structured recovery instruction
// =============================================================================

#[test]
fn test_truncated_tool_call_includes_structured_recovery() {
    // @step Given the agent is streaming a response from any provider
    // @step And the model attempts a Write tool call with content exceeding the output token limit
    let error = "Streaming error: ResponseError: Tool call truncated due to output token limit. \
                 Tool 'Write' received incomplete JSON arguments. \
                 The model hit max_tokens while generating the tool call. \
                 Partial arguments: {\"file_path\": \"/tmp/prov039-large-write-test.txt\"";

    // @step When the tool call is truncated due to max_tokens
    assert!(
        is_truncated_tool_call_error(error),
        "Should detect truncated tool call error"
    );

    // @step Then the error message contains a structured recovery instruction
    let recovery = build_truncation_recovery_message(error);
    assert!(!recovery.is_empty(), "Recovery message should not be empty");

    // @step And the recovery instruction suggests using Bash with heredoc for large files
    assert!(
        recovery.contains("Bash") && recovery.contains("heredoc"),
        "Recovery message should suggest Bash with heredoc. Got: {recovery}"
    );

    // @step And the recovery instruction suggests splitting into multiple smaller Write calls
    assert!(
        recovery.contains("smaller Write"),
        "Recovery message should suggest splitting into smaller writes. Got: {recovery}"
    );

    // @step And the recovery instruction includes the truncated tool name
    assert!(
        recovery.contains("Your Write tool call was truncated"),
        "Recovery message should include tool name 'Write'. Got: {recovery}"
    );

    // @step And the recovery instruction includes the partial arguments that were received
    assert!(
        recovery.contains("/tmp/prov039-large-write-test.txt"),
        "Recovery message should include partial arguments. Got: {recovery}"
    );
}

// =============================================================================
// Scenario: Retry budget prevents infinite truncation retry loops
// =============================================================================

#[test]
fn test_retry_budget_prevents_infinite_loops() {
    // @step Given the agent is streaming a response from any provider
    let error = "Tool call truncated due to output token limit. Tool 'Write' received incomplete JSON arguments. Partial arguments: {}";

    // @step And the truncation retry budget is set to 2
    // Verify the production constant matches the expected budget
    assert_eq!(
        MAX_TRUNCATION_RETRIES, 2,
        "Production retry budget must be 2"
    );

    // @step And the model has already exhausted the retry budget with consecutive truncation errors
    // Simulate the retry counter logic from stream_loop.rs lines 1547-1549:
    //   truncation_retry_count += 1;
    //   if truncation_retry_count <= MAX_TRUNCATION_RETRIES { /* retry */ }
    // First two attempts are within budget:
    let mut truncation_retry_count: u32 = 0;
    for attempt in 1..=MAX_TRUNCATION_RETRIES {
        assert!(is_truncated_tool_call_error(error));
        truncation_retry_count += 1;
        assert!(
            truncation_retry_count <= MAX_TRUNCATION_RETRIES,
            "Attempt {} should be within budget (count={}, max={})",
            attempt,
            truncation_retry_count,
            MAX_TRUNCATION_RETRIES
        );
        // In production: recovery message is built and retry stream is started
        let recovery = build_truncation_recovery_message(error);
        assert!(!recovery.is_empty(), "Recovery should be generated for attempt {}", attempt);
    }

    // Third attempt exceeds the budget:
    assert!(is_truncated_tool_call_error(error));
    truncation_retry_count += 1;
    assert!(
        truncation_retry_count > MAX_TRUNCATION_RETRIES,
        "Attempt {} should exceed budget (count={}, max={})",
        truncation_retry_count,
        truncation_retry_count,
        MAX_TRUNCATION_RETRIES
    );

    // @step When the budget-exhausted error is generated
    let budget_error = build_truncation_budget_exhausted_message(MAX_TRUNCATION_RETRIES);

    // @step Then the error message includes the retry count and informs the user the budget is exhausted
    assert!(
        budget_error.contains(&MAX_TRUNCATION_RETRIES.to_string()),
        "Budget error should include the retry count. Got: {budget_error}"
    );
    assert!(
        budget_error.contains("retry budget exhausted"),
        "Budget error should inform the user the budget is exhausted. Got: {budget_error}"
    );

    // @step And the error message suggests alternative strategies for large content
    assert!(
        budget_error.contains("Bash") && budget_error.contains("heredoc"),
        "Budget error should suggest Bash with heredoc. Got: {budget_error}"
    );
    assert!(
        budget_error.contains("smaller operations"),
        "Budget error should suggest splitting into smaller operations. Got: {budget_error}"
    );

    // @step And the stream loop terminates without starting another retry
    // In production (stream_loop.rs:1642), this returns Err(anyhow!(...))
    // which terminates the loop. We verify the termination condition holds:
    assert!(
        truncation_retry_count > MAX_TRUNCATION_RETRIES,
        "After budget exhaustion, the retry count ({}) must exceed the max ({}), preventing further retries",
        truncation_retry_count,
        MAX_TRUNCATION_RETRIES
    );
}

// =============================================================================
// Scenario: Normal completion is unaffected by truncation recovery logic
// =============================================================================

#[test]
fn test_normal_completion_unaffected() {
    // @step Given the agent is streaming a response from any provider
    // @step And the model completes a tool call normally with stop_reason end_turn
    // These represent errors that occur during normal operation (not truncation)
    // including end_turn completions that may carry status messages
    let normal_errors = [
        "Network timeout",
        "Authentication failed",
        "Rate limit exceeded",
        "prompt is too long: 209834 tokens > 200000 maximum",
        "context_length_exceeded: Request too large",
        "SSE Error: connection reset",
    ];

    // @step When the stream completes
    // Track what a truncation_retry_count variable would do (stays at zero)
    let mut truncation_retry_count: u32 = 0;

    for error in &normal_errors {
        // @step Then no recovery instruction is injected
        let is_truncation = is_truncated_tool_call_error(error);
        assert!(
            !is_truncation,
            "Normal error should NOT trigger truncation recovery: {error}"
        );
        // In production, truncation_retry_count is only incremented inside
        // the `if is_truncated_tool_call_error(...)` block (stream_loop.rs:1547)
        if is_truncation {
            truncation_retry_count += 1;
        }
    }

    // @step And the truncation retry counter remains at zero
    assert_eq!(
        truncation_retry_count, 0,
        "Truncation retry counter must remain at zero for non-truncation errors"
    );

    // @step And the behavior is identical to pre-PROV-040 baseline
    // Pre-PROV-040: errors flow straight through to the error handler
    // Post-PROV-040: the `is_truncated_tool_call_error` check is the ONLY new gate
    // Since it returns false for all normal errors, the code path is identical.
    // Verify none of these normal errors would generate a recovery message:
    for error in &normal_errors {
        // build_truncation_recovery_message would generate a message even for non-truncation
        // errors (it just does string parsing), but it's NEVER called because the
        // is_truncated_tool_call_error gate blocks it. Verify that gate is closed:
        assert!(
            !is_truncated_tool_call_error(error),
            "Pre-PROV-040 error path must be preserved for: {error}"
        );
    }
}

// =============================================================================
// Scenario: Text-only truncation does not trigger tool call recovery
// =============================================================================

#[test]
fn test_text_truncation_no_recovery() {
    // @step Given the agent is streaming a response from any provider
    // @step And the model hits max_tokens during a text-only response with no tool call
    // Text-only truncation produces a different error message than tool call truncation.
    // PROV-039 only enriches the error with "Tool call truncated due to output token limit"
    // when there is a pending (unclosed) tool call at the time max_tokens fires.
    let text_truncation_messages = [
        "Response truncated: model hit max_tokens output limit",
        "max_tokens",
        "Output limit reached",
    ];

    // @step When the stream completes with stop_reason max_tokens
    for msg in &text_truncation_messages {
        // @step Then the existing PROV-039 truncation warning is displayed
        // In production, text-only truncation is handled by the StreamEvent::Done(Some("max_tokens"))
        // branch in stream_loop.rs, which emits a warning via output.emit_warning().
        // The key invariant we verify here is that the PROV-040 recovery path is NOT entered:
        assert!(
            !is_truncated_tool_call_error(msg),
            "Text-only truncation message should NOT trigger tool call recovery: {msg}"
        );

        // @step And no tool call recovery instruction is injected
        // Since is_truncated_tool_call_error returns false, the recovery branch
        // at stream_loop.rs:1546 is never entered, and build_truncation_recovery_message
        // is never called. The PROV-039 text truncation warning flows through separately.
    }
}

// =============================================================================
// Scenario: Truncation recovery is provider-agnostic
// =============================================================================

#[test]
fn test_provider_agnostic_detection() {
    // @step Given the truncation detection relies on the error message string from PROV-039
    let anthropic_error =
        "Streaming error: ResponseError: Tool call truncated due to output token limit. \
         Tool 'Write' received incomplete JSON arguments. \
         Partial arguments: {\"file_path\": \"/tmp/test.txt\"";
    let openai_error =
        "Tool call truncated due to output token limit. \
         Tool 'Edit' received incomplete JSON arguments. \
         Partial arguments: {\"file_path\": \"/src/main.rs\"";
    let gemini_error =
        "Tool call truncated due to output token limit. \
         Tool 'Bash' received incomplete JSON arguments. \
         Partial arguments: {\"command\": \"echo test\"";

    // @step When a truncation error containing "Tool call truncated due to output token limit" is received
    // @step Then the same recovery logic fires regardless of whether the provider is Anthropic, OpenAI, or Gemini
    assert!(is_truncated_tool_call_error(anthropic_error), "Should detect Anthropic truncation");
    assert!(is_truncated_tool_call_error(openai_error), "Should detect OpenAI truncation");
    assert!(is_truncated_tool_call_error(gemini_error), "Should detect Gemini truncation");

    // @step And the recovery instruction content is identical across all providers
    let recovery_anthropic = build_truncation_recovery_message(anthropic_error);
    let recovery_openai = build_truncation_recovery_message(openai_error);
    let recovery_gemini = build_truncation_recovery_message(gemini_error);

    // All recovery messages must contain the same strategy suggestions
    for recovery in [&recovery_anthropic, &recovery_openai, &recovery_gemini] {
        assert!(recovery.contains("Bash"), "All recoveries should suggest Bash");
        assert!(recovery.contains("heredoc"), "All recoveries should suggest heredoc");
        assert!(recovery.contains("smaller Write"), "All recoveries should suggest splitting");
    }

    // Recovery messages differ only in tool name and partial args — verify tool names are correct
    assert!(recovery_anthropic.contains("Your Write tool call was truncated"));
    assert!(recovery_openai.contains("Your Edit tool call was truncated"));
    assert!(recovery_gemini.contains("Your Bash tool call was truncated"));
}

// =============================================================================
// Edge case tests — additional coverage beyond scenarios
// =============================================================================

#[test]
fn test_empty_error_string_not_detected() {
    assert!(!is_truncated_tool_call_error(""));
}

#[test]
fn test_partial_match_not_detected() {
    assert!(!is_truncated_tool_call_error("Tool call truncated"));
    assert!(!is_truncated_tool_call_error("output token limit"));
}

#[test]
fn test_detection_is_case_sensitive() {
    assert!(!is_truncated_tool_call_error("tool call truncated due to output token limit"));
    assert!(is_truncated_tool_call_error("Tool call truncated due to output token limit"));
}

#[test]
fn test_recovery_message_for_bash_tool() {
    let error = "Tool call truncated due to output token limit. \
                 Tool 'Bash' received incomplete JSON arguments. \
                 Partial arguments: {\"command\": \"cat << 'EOF'";
    let recovery = build_truncation_recovery_message(error);
    assert!(recovery.contains("Your Bash tool call was truncated"));
    assert!(recovery.contains("Do NOT retry the same large Bash call"));
}

#[test]
fn test_recovery_message_handles_missing_tool_name() {
    let error = "Tool call truncated due to output token limit. No standard format here.";
    let recovery = build_truncation_recovery_message(error);
    assert!(recovery.contains("unknown"));
}

#[test]
fn test_recovery_message_handles_missing_partial_args() {
    let error = "Tool call truncated due to output token limit. \
                 Tool 'Write' received incomplete JSON arguments.";
    let recovery = build_truncation_recovery_message(error);
    assert!(recovery.contains("(not available)"));
}

#[test]
fn test_recovery_message_warns_against_retry() {
    let error = "Tool call truncated due to output token limit. \
                 Tool 'Write' received incomplete JSON arguments. \
                 Partial arguments: {\"file_path\": \"/tmp/big.txt\"";
    let recovery = build_truncation_recovery_message(error);
    assert!(
        recovery.contains("Do NOT retry the same large Write call"),
        "Recovery should explicitly warn against retry. Got: {recovery}"
    );
}

#[test]
fn test_budget_exhausted_message_content() {
    // Verify the budget-exhausted message contains all required components
    let msg = build_truncation_budget_exhausted_message(MAX_TRUNCATION_RETRIES);
    assert!(msg.contains("2"), "Should include retry count");
    assert!(msg.contains("retry budget exhausted"), "Should state budget exhaustion");
    assert!(msg.contains("Bash"), "Should suggest Bash alternative");
    assert!(msg.contains("heredoc"), "Should suggest heredoc");
    assert!(msg.contains("smaller operations"), "Should suggest splitting");
}

#[test]
fn test_budget_exhausted_message_with_custom_count() {
    // Verify the message correctly formats with different retry counts
    let msg_1 = build_truncation_budget_exhausted_message(1);
    assert!(msg_1.contains("1 times"), "Should show count of 1");
    let msg_5 = build_truncation_budget_exhausted_message(5);
    assert!(msg_5.contains("5 times"), "Should show count of 5");
}
